File: mod/Recipe/Phase.cs
namespace USS {
  namespace Recipe {
    public class Phase(string origin, string type, float temp, (string, float)[] dest): IRecipe {
      public string origin = origin;
      public string type = type;
      public float temp = temp;
      public Dictionary<string, float> dest = dest.ToDictionary(d => d.Item1, d => d.Item2);

      public void ToRust(StreamWriter file)
      {
        file.WriteLine("Recipe::Phase(Phase {");
        file.WriteLine("input: &" + origin + ",");
        file.WriteLine("temp: crate::units::Temperature(" + temp.ToString("0.000") + "),");
        file.WriteLine("type_of: PhaseType::" + type[0].ToString().ToUpper() + type.Substring(1) + ",");
        file.WriteLine("output: &[");
        foreach (var item in dest)
        {
          file.WriteLine("EntityPercent{entity: &" + item.Key + ", percent: crate::units::Percent(" + item.Value.ToString("0.000") + ")},");
        }
        file.WriteLine("]");
        file.WriteLine("}),");
      }

      public static void ToRustClass(StreamWriter file)
      {
        file.WriteLine("""
pub enum PhaseType {
  Cooking,
  Melting,
  Solidification,
  Vaporization,
  Freezing,
  Sublimation,
  Condensation,
  Transmutation,
}
pub struct Phase {
  pub input: &'static Entity,
  pub temp: crate::units::Temperature,
  pub type_of: PhaseType,
  pub output: &'static[EntityPercent],
}              
""");
      }
    }
  }
}

File: src/html/recipe/phase.rs
use crate::db::recipe::*;
use crate::html::*;
use crate::icon;

pub fn display_phases(container: &HtmlElement, entity: &Entity) {
  let align = Align::new();
  for recipe in recipes.iter().filter_map(|r| match r {
    Recipe::Phase(ret @ Phase { input: other, .. }) if std::ptr::eq(entity, *other) => Some(ret),
    _ => None,
  }) {
    let icon = match recipe.type_of {
      PhaseType::Cooking => &icon::TEMP_UP,
      PhaseType::Melting => &icon::TEMP_UP,
      PhaseType::Solidification => &icon::TEMP_UP,
      PhaseType::Vaporization => &icon::TEMP_UP,
      PhaseType::Freezing => &icon::TEMP_DOWN,
      PhaseType::Sublimation => &icon::TEMP_UP,
      PhaseType::Transmutation => &icon::TEMP_UP,
      PhaseType::Condensation => &icon::TEMP_DOWN,
    };
    Line::new(container).icon(icon).label(recipe.type_of.to_text()).values("propValuePhase", |line| {
      line
        .value_align("valuePhaseTemp", &recipe.temp, &align)
        .value("valuePhaseArrow", "⟶")
        .foreach(recipe.output, |line, output| {
          line
            .img("valuePhaseIcon", &output.entity.img())
            .value("valuePhaseElement", output.entity.name())
            .value_filter("valuePhasePercent", &output.percent, output.percent.0 < 1.0)
        })
    });
  }
}

pub fn get_other_phases(entity: &Entity) -> Vec<&Phase>{
  recipes.iter().filter_map(|r| match r {
    Recipe::Phase(ret @ Phase { output: other, .. }) if other.iter().any(|o| std::ptr::eq(entity, o.entity)) => Some(ret),
    _ => None,
  }).collect::<Vec<_>>()
}

pub fn display_other_phases(container: &HtmlElement, other_phases: Vec<&Phase>) {
  let align = Align::new();
  for recipe in other_phases {
    let icon = match recipe.type_of {
      PhaseType::Cooking => &icon::TEMP_UP,
      PhaseType::Melting => &icon::TEMP_UP,
      PhaseType::Solidification => &icon::TEMP_UP,
      PhaseType::Vaporization => &icon::TEMP_UP,
      PhaseType::Freezing => &icon::TEMP_DOWN,
      PhaseType::Sublimation => &icon::TEMP_UP,
      PhaseType::Condensation => &icon::TEMP_DOWN,
      PhaseType::Transmutation => &icon::TEMP_UP,
    };
    Line::new(container).icon(&recipe.input.img()).label(recipe.input.name()).values("propValuePhase", |line| {
      line
        .img("valuePhaseTempIcon", icon)
        .value_align("valuePhaseTemp", &recipe.temp, &align)
        .value("valuePhaseArrow", "⟶")
        .foreach(recipe.output, |line, output| {
          line
            .img("valuePhaseIcon", &output.entity.img())
            .value("valuePhaseElement", output.entity.name())
            .value_filter("valuePhasePercent", &output.percent, output.percent.0 < 1.0)
        })
    });
  }
}
